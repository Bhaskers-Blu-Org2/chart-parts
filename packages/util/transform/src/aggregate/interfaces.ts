import { OperatorFunction } from 'rxjs'
import { FieldAccessor } from '../interfaces'

/**
 * Types of aggregate field operations
 */
export enum AggregateOperation {
	/**
	 * The total count of data objects in the group.
	 */
	count = 'count',
	/**
	 * The count of field values that are not null, undefined or NaN.
	 */
	valid = 'valid',

	/**
	 * The count of null or undefined field values.
	 */
	missing = 'missing',

	/**
	 * The count of distinct field values.
	 */
	distinct = 'distinct',

	/**
	 * The sum of field values.
	 */
	sum = 'sum',

	/**
	 * The mean (average) field value.
	 */
	mean = 'mean',

	/**
	 * The sample variance of field values.
	 */
	variance = 'variance',

	/**
	 * The sample standard deviation of field values.
	 */
	stdev = 'stdev',

	/**
	 * 	The median field value.
	 */
	median = 'median',

	/**
	 * The lower quartile boundary of field values.
	 */
	q1 = 'q1',

	/**
	 * The upper quartile boundary of field values.
	 */
	q3 = 'q3',

	/**
	 * The minimum field value.
	 */
	min = 'min',

	/**
	 * The maximum field value.
	 */
	max = 'max',
}

/**
 * Describes an aggregate computation field. This indicates performing a calculation over a field in a dataset as a new output field
 */
export interface AggregateComputeField {
	field: FieldAccessor
	op: AggregateOperation
	as: string
}

export interface AggregateBuilder<T> extends OperatorFunction<T[], any> {
	groupBy(...fields: FieldAccessor[]): AggregateBuilder<T>
	compute(...fields: AggregateComputeField[]): AggregateBuilder<T>
}
