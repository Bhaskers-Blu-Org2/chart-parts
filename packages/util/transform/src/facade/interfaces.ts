export interface DatasetTransform {
	build(df: any, from: any): any
}
