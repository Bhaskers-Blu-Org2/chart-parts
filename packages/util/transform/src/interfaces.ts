export type FieldAccessor = string // | ((row: any) => number)

export enum CompareOrder {
	ascending = 'asc',
	descending = 'desc',
}

export interface Compare {
	field: FieldAccessor
	order?: CompareOrder
}

export enum Offset {
	zero = 'zero',
	center = 'center',
	normalize = 'normalize',
}

export type CompareFunction<T> = (a: T, b: T) => number
export type EqualityFunction<T> = (a: T, b: T) => boolean
export type GetKeyFunction<T> = (item: T) => any
export type Predicate<T> = (item: T) => boolean
export type Transformer<T, K> = (input: T) => K

export type MaybeNumber = number | undefined
