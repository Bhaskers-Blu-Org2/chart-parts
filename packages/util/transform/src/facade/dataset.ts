// tslint:disable no-var-requires
import { DatasetTransform } from './interfaces'
declare var require: any
const { Dataflow, changeset } = require('vega-dataflow')
const { collect } = require('vega-transforms')

/**
 * Storage structure for tracking datasets
 */
interface Pipeline {
	dataflow: any
	start: any
	end: any
}

export interface DatasetManager {
	/**
	 * Adds a new data table to the dataset
	 * @param name the name of the dataset to add
	 * @param dataset the array of data rows to add
	 */
	add(
		name: string,
		dataset: any[],
		...transforms: DatasetTransform[]
	): DatasetManager

	/**
	 * Gets a dataset
	 * @param name The name of the dataset
	 */
	get(name: string): any[] | undefined
}

export class DatasetManagerImpl implements DatasetManager {
	// Datasets are stored in a vega transform pipeline
	private pipelines: Map<string, Pipeline> = new Map()

	public add(name: string, data: any[], ...transforms: DatasetTransform[]) {
		const pipeline = createTransformPipeline(transforms)
		pushData(data, pipeline)
		this.pipelines.set(name, pipeline)
		return this
	}

	public get(name: string): any[] | undefined {
		const ds = this.pipelines.get(name)
		return ds && ds.end.value
	}
}

function pushData(data: any[], pipeline: Pipeline) {
	const { dataflow, start } = pipeline
	dataflow.pulse(start, changeset().insert(data)).run()
}

function createTransformPipeline(transforms: DatasetTransform[]): Pipeline {
	const df = new Dataflow()
	const entry = df.add(collect)
	let latest = entry

	transforms.forEach(t => (latest = t.build(df, latest)))
	latest = df.add(collect, { pulse: latest })

	return {
		dataflow: df,
		start: entry,
		end: latest,
	}
}

/**
 * Create a new dataset builder
 */
export function dataset(): DatasetManager {
	return new DatasetManagerImpl()
}
