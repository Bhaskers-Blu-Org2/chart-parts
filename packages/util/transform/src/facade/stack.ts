// tslint:disable no-var-requires
import { FieldAccessor, Compare, Offset } from '../interfaces'
import { DatasetTransform } from './interfaces'
import { createSorter } from '../util'

declare var require: any
const { stack: vegaStack } = require('vega-encode')
const { field } = require('vega-util')

export interface StackBuilder extends DatasetTransform {
	groupBy(...fields: FieldAccessor[]): StackBuilder
	sort(compare: Compare): StackBuilder
	offset(offset: Offset): StackBuilder
	as(start: string, end: string): StackBuilder
}

export class StackBuilderImpl implements StackBuilder {
	private groupByFields: FieldAccessor[] | undefined
	private compareValue: Compare | undefined
	private offsetValue: Offset | undefined
	private asValue: [string, string] | undefined

	constructor(private stackField: string) {}

	public groupBy(...fields: FieldAccessor[]) {
		this.groupByFields = fields
		return this
	}

	public sort(compare: Compare) {
		this.compareValue = compare
		return this
	}

	public offset(offset: Offset) {
		this.offsetValue = offset
		return this
	}

	public as(start: string, end: string) {
		this.asValue = [start, end]
		return this
	}

	public build(df: any, from: any) {
		const spec: any = {
			field: field(this.stackField),
			offset: this.offsetValue,
			pulse: from,
		}
		if (this.groupByFields) {
			spec.groupby = this.groupByFields.map(c => field(c))
		}
		if (this.compareValue) {
			spec.sort = createSorter(this.compareValue)
		}
		if (this.asValue) {
			spec.as = this.asValue
		}

		const stackNode = df.add(vegaStack, spec)
		return stackNode
	}
}

export function stack(stackField: FieldAccessor): StackBuilder {
	return new StackBuilderImpl(stackField)
}
