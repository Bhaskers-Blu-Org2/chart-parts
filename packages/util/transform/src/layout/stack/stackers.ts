import { getField } from '../../util'
import { FieldAccessor, Offset } from '../../interfaces'
import { StackBuilderContext } from './index'

export type Stacker = (
	context: StackBuilderContext,
	rows: any[],
	sum: number,
	max: number,
) => void

export function getStacker(context: StackBuilderContext): Stacker {
	switch (context.offset) {
		case Offset.zero:
			return stackZero
		case Offset.normalize:
			return stackNormalize
		case Offset.center:
			return stackCenter
		default:
			throw new Error('could not handle stack offset: ' + context.offset)
	}
}

const stackCenter = (
	context: StackBuilderContext,
	rows: any[],
	sum: number,
	max: number,
) => {
	const [y0, y1] = context.outputFields
	const field = context.field as FieldAccessor
	let last = (max - sum) / 2
	for (const d of rows) {
		const value = Math.abs(getField(d, field))
		d[y0] = last
		d[y1] = last += value
	}
}

const stackNormalize = (
	context: StackBuilderContext,
	rows: any[],
	sum: number,
) => {
	const [y0, y1] = context.outputFields
	const field = context.field
	if (!field) {
		throw new Error('StackTransform: field must be defined')
	}
	const scale = 1 / sum

	let last = 0
	let v = 0
	for (const d of rows) {
		const value = Math.abs(getField(d, field))
		d[y0] = last
		d[y1] = last = scale * (v += value)
	}
}

const stackZero = (context: StackBuilderContext, rows: any[]) => {
	const [y0, y1] = context.outputFields
	const field = context.field

	if (!field) {
		throw new Error('StackTransform: field must be defined')
	}

	let lastPos = 0
	let lastNeg = 0
	for (const d of rows) {
		const v = getField(d, field)
		if (v < 0) {
			d[y0] = lastNeg
			d[y1] = lastNeg += v
		} else {
			d[y0] = lastPos
			d[y1] = lastPos += v
		}
	}
}
