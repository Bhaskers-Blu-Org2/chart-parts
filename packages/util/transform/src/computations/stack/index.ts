import {
	Observable,
	OperatorFunction,
	Subscriber,
	isObservable,
	zip,
} from 'rxjs'
import { takeLast, toArray } from 'rxjs/operators'
import { FieldAccessor, Compare, Offset } from '../../interfaces'
import max from '../max'
import sum from '../sum'
import accessField from '../accessField'
import { getStacker } from './stackers'

export interface StackBuilder extends OperatorFunction<Observable<any>, any> {
	sort(...sorts: Compare[])
	offset(value: Offset)
	output(start: string, end: string)
}

export interface StackBuilderContext {
	sorts: Compare[]
	offset: Offset
	field: FieldAccessor
	outputFields: [string, string]
}

/**
 * Creates an observable node based on incoming number stream
 * @param source An observable of numbers to emit the maximum value of
 */
export default function stack(field: FieldAccessor) {
	const context: StackBuilderContext = {
		field,
		sorts: [],
		offset: Offset.zero,
		outputFields: ['y0', 'y1'],
	}
	function makeStack(source: Observable<Observable<any>>) {
		return Observable.create((subscriber: Subscriber<any>) =>
			source.subscribe(
				group => {
					if (!isObservable(group)) {
						subscriber.error(
							new Error(
								`stack group is not an observable. 
                            You may have forgotten to apply a groupBy() before applying stack()
                            `,
							),
						)
					}
					try {
						processGroup(group, context, outputRow =>
							subscriber.next(outputRow),
						)
					} catch (err) {
						subscriber.error(err)
					}
				},
				err => subscriber.error(err),
				() => subscriber.complete(),
			),
		)
	}
	;(makeStack as any).sort = (...sortBy: Compare[]) => {
		context.sorts = sortBy
		return makeStack
	}
	;(makeStack as any).offset = (value: Offset) => {
		context.offset = value
		return makeStack
	}
	;(makeStack as any).output = (start: string, end: string) => {
		context.outputFields = [start, end]
		return makeStack
	}
	return makeStack as StackBuilder
}

function processGroup(
	group: Observable<any>,
	context: StackBuilderContext,
	handleOutputRow: (row: any) => void,
) {
	const groupMax = group.pipe(
		accessField<number>(context.field),
		max(),
		takeLast(1),
	)
	const groupSum = group.pipe(
		accessField<number>(context.field),
		sum(),
		takeLast(1),
	)
	const groupItems = group.pipe(toArray())
	zip(groupMax, groupSum, groupItems).subscribe(
		([maxValue, sumValue, rows]) => {
			// Apply the stack logic
			getStacker(context)(context, rows, sumValue, maxValue)
			rows.forEach(row => handleOutputRow(row))
		},
	)
}
