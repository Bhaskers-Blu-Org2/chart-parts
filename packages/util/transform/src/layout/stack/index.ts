import { Observable, OperatorFunction } from 'rxjs'
import { max, sum } from 'd3-array'
import { FieldAccessor, Compare, Offset } from '../../interfaces'
import { createOperator, getField, flatMap } from '../../util'
import { groupBy, GroupedData } from '../../groupBy'
import { getStacker } from './stackers'

export interface StackBuilder<T> extends OperatorFunction<T[], any> {
	sort(...sorts: Compare[]): StackBuilder<T>
	offset(value: Offset): StackBuilder<T>
	output(start: string, end: string): StackBuilder<T>
	groupBy(...fields: FieldAccessor[]): StackBuilder<T>
}

export interface StackBuilderContext {
	sorts: Compare[]
	groupBy: FieldAccessor[]
	offset: Offset
	field: FieldAccessor
	outputFields: [string, string]
}

/**
 * Creates an observable node based on incoming number stream
 * @param source An observable of numbers to emit the maximum value of
 */
export function stack<T>(field: FieldAccessor): StackBuilder<T> {
	const context: StackBuilderContext = {
		field,
		groupBy: [],
		sorts: [],
		offset: Offset.zero,
		outputFields: ['y0', 'y1'],
	}

	const getFieldValue = (d: any) => getField(d, context.field)

	function transform(data: GroupedData<T>) {
		const { groups } = data
		function processStacks() {
			const keys = Array.from(groups.keys())
			return keys.map(groupKey => {
				const rows = Array.from(groups.get(groupKey) as any)
				const maxValue = max(rows, getFieldValue)
				const sumValue = sum(rows, getFieldValue)
				const stacker = getStacker(context)
				return stacker(context, rows, sumValue, maxValue as number)
			})
		}

		return flatMap(processStacks(), (t: any) => t)
	}

	const makeStack = createOperator({
		transform,
		handleSource: (input: Observable<T[]>) => {
			return input.pipe(groupBy(...context.groupBy))
		},
	})

	const builder = (makeStack as any) as StackBuilder<T>
	builder.sort = (...sortBy: Compare[]) => {
		context.sorts = sortBy
		return builder
	}
	builder.groupBy = (...fields: FieldAccessor[]) => {
		context.groupBy = fields
		return builder
	}
	builder.offset = (value: Offset) => {
		context.offset = value
		return builder
	}
	builder.output = (start: string, end: string) => {
		context.outputFields = [start, end]
		return builder
	}
	return builder
}
