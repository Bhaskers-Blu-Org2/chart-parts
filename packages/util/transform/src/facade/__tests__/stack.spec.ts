import { stack } from '../stack'
import { dataset } from '../dataset'
import { Offset, CompareOrder } from '../../interfaces'

const data = [
	{ x: 0, y: 28, c: 0 },
	{ x: 0, y: 55, c: 1 },
	{ x: 1, y: 43, c: 0 },
	{ x: 1, y: 91, c: 1 },
	{ x: 2, y: 81, c: 0 },
	{ x: 2, y: 53, c: 1 },
	{ x: 3, y: 19, c: 0 },
	{ x: 3, y: 87, c: 1 },
	{ x: 4, y: 52, c: 0 },
	{ x: 4, y: 48, c: 1 },
	{ x: 5, y: 24, c: 0 },
	{ x: 5, y: 49, c: 1 },
	{ x: 6, y: 87, c: 0 },
	{ x: 6, y: 66, c: 1 },
	{ x: 7, y: 17, c: 0 },
	{ x: 7, y: 27, c: 1 },
	{ x: 8, y: 68, c: 0 },
	{ x: 8, y: 16, c: 1 },
	{ x: 9, y: 49, c: 0 },
	{ x: 9, y: 15, c: 1 },
]

describe('The stack transformation', () => {
	it('can stack values', () => {
		const ds = dataset()
			.add('data', data, stack('y'))
			.add('data-grouped', data, stack('y').groupBy('x'))
			.add(
				'data-grouped-sorted-and-offset',
				data,
				stack('y')
					.groupBy('x')
					.sort({ field: 'c', order: CompareOrder.descending })
					.offset(Offset.center),
			)

		// The result should not be the same instance (i.e. preserve the input for functional style)
		expect(ds.get('data')).toMatchSnapshot()
		expect(ds.get('data-grouped')).toMatchSnapshot()
		expect(ds.get('data-grouped-sorted-and-offset')).toMatchSnapshot()
	})
})
