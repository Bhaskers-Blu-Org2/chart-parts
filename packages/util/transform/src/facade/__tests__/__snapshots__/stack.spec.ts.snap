// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The stack transformation can stack values 1`] = `
Array [
  Object {
    "c": 0,
    "x": 0,
    "y": 28,
    "y0": 35,
    "y1": 63,
    Symbol(vega_id): 1,
  },
  Object {
    "c": 1,
    "x": 0,
    "y": 55,
    "y0": 63,
    "y1": 118,
    Symbol(vega_id): 2,
  },
  Object {
    "c": 0,
    "x": 1,
    "y": 43,
    "y0": 9.5,
    "y1": 52.5,
    Symbol(vega_id): 3,
  },
  Object {
    "c": 1,
    "x": 1,
    "y": 91,
    "y0": 52.5,
    "y1": 143.5,
    Symbol(vega_id): 4,
  },
  Object {
    "c": 0,
    "x": 2,
    "y": 81,
    "y0": 9.5,
    "y1": 90.5,
    Symbol(vega_id): 5,
  },
  Object {
    "c": 1,
    "x": 2,
    "y": 53,
    "y0": 90.5,
    "y1": 143.5,
    Symbol(vega_id): 6,
  },
  Object {
    "c": 0,
    "x": 3,
    "y": 19,
    "y0": 23.5,
    "y1": 42.5,
    Symbol(vega_id): 7,
  },
  Object {
    "c": 1,
    "x": 3,
    "y": 87,
    "y0": 42.5,
    "y1": 129.5,
    Symbol(vega_id): 8,
  },
  Object {
    "c": 0,
    "x": 4,
    "y": 52,
    "y0": 26.5,
    "y1": 78.5,
    Symbol(vega_id): 9,
  },
  Object {
    "c": 1,
    "x": 4,
    "y": 48,
    "y0": 78.5,
    "y1": 126.5,
    Symbol(vega_id): 10,
  },
  Object {
    "c": 0,
    "x": 5,
    "y": 24,
    "y0": 40,
    "y1": 64,
    Symbol(vega_id): 11,
  },
  Object {
    "c": 1,
    "x": 5,
    "y": 49,
    "y0": 64,
    "y1": 113,
    Symbol(vega_id): 12,
  },
  Object {
    "c": 0,
    "x": 6,
    "y": 87,
    "y0": 0,
    "y1": 87,
    Symbol(vega_id): 13,
  },
  Object {
    "c": 1,
    "x": 6,
    "y": 66,
    "y0": 87,
    "y1": 153,
    Symbol(vega_id): 14,
  },
  Object {
    "c": 0,
    "x": 7,
    "y": 17,
    "y0": 54.5,
    "y1": 71.5,
    Symbol(vega_id): 15,
  },
  Object {
    "c": 1,
    "x": 7,
    "y": 27,
    "y0": 71.5,
    "y1": 98.5,
    Symbol(vega_id): 16,
  },
  Object {
    "c": 0,
    "x": 8,
    "y": 68,
    "y0": 34.5,
    "y1": 102.5,
    Symbol(vega_id): 17,
  },
  Object {
    "c": 1,
    "x": 8,
    "y": 16,
    "y0": 102.5,
    "y1": 118.5,
    Symbol(vega_id): 18,
  },
  Object {
    "c": 0,
    "x": 9,
    "y": 49,
    "y0": 44.5,
    "y1": 93.5,
    Symbol(vega_id): 19,
  },
  Object {
    "c": 1,
    "x": 9,
    "y": 15,
    "y0": 93.5,
    "y1": 108.5,
    Symbol(vega_id): 20,
  },
]
`;

exports[`The stack transformation can stack values 2`] = `
Array [
  Object {
    "c": 0,
    "x": 0,
    "y": 28,
    "y0": 35,
    "y1": 63,
    Symbol(vega_id): 1,
  },
  Object {
    "c": 1,
    "x": 0,
    "y": 55,
    "y0": 63,
    "y1": 118,
    Symbol(vega_id): 2,
  },
  Object {
    "c": 0,
    "x": 1,
    "y": 43,
    "y0": 9.5,
    "y1": 52.5,
    Symbol(vega_id): 3,
  },
  Object {
    "c": 1,
    "x": 1,
    "y": 91,
    "y0": 52.5,
    "y1": 143.5,
    Symbol(vega_id): 4,
  },
  Object {
    "c": 0,
    "x": 2,
    "y": 81,
    "y0": 9.5,
    "y1": 90.5,
    Symbol(vega_id): 5,
  },
  Object {
    "c": 1,
    "x": 2,
    "y": 53,
    "y0": 90.5,
    "y1": 143.5,
    Symbol(vega_id): 6,
  },
  Object {
    "c": 0,
    "x": 3,
    "y": 19,
    "y0": 23.5,
    "y1": 42.5,
    Symbol(vega_id): 7,
  },
  Object {
    "c": 1,
    "x": 3,
    "y": 87,
    "y0": 42.5,
    "y1": 129.5,
    Symbol(vega_id): 8,
  },
  Object {
    "c": 0,
    "x": 4,
    "y": 52,
    "y0": 26.5,
    "y1": 78.5,
    Symbol(vega_id): 9,
  },
  Object {
    "c": 1,
    "x": 4,
    "y": 48,
    "y0": 78.5,
    "y1": 126.5,
    Symbol(vega_id): 10,
  },
  Object {
    "c": 0,
    "x": 5,
    "y": 24,
    "y0": 40,
    "y1": 64,
    Symbol(vega_id): 11,
  },
  Object {
    "c": 1,
    "x": 5,
    "y": 49,
    "y0": 64,
    "y1": 113,
    Symbol(vega_id): 12,
  },
  Object {
    "c": 0,
    "x": 6,
    "y": 87,
    "y0": 0,
    "y1": 87,
    Symbol(vega_id): 13,
  },
  Object {
    "c": 1,
    "x": 6,
    "y": 66,
    "y0": 87,
    "y1": 153,
    Symbol(vega_id): 14,
  },
  Object {
    "c": 0,
    "x": 7,
    "y": 17,
    "y0": 54.5,
    "y1": 71.5,
    Symbol(vega_id): 15,
  },
  Object {
    "c": 1,
    "x": 7,
    "y": 27,
    "y0": 71.5,
    "y1": 98.5,
    Symbol(vega_id): 16,
  },
  Object {
    "c": 0,
    "x": 8,
    "y": 68,
    "y0": 34.5,
    "y1": 102.5,
    Symbol(vega_id): 17,
  },
  Object {
    "c": 1,
    "x": 8,
    "y": 16,
    "y0": 102.5,
    "y1": 118.5,
    Symbol(vega_id): 18,
  },
  Object {
    "c": 0,
    "x": 9,
    "y": 49,
    "y0": 44.5,
    "y1": 93.5,
    Symbol(vega_id): 19,
  },
  Object {
    "c": 1,
    "x": 9,
    "y": 15,
    "y0": 93.5,
    "y1": 108.5,
    Symbol(vega_id): 20,
  },
]
`;

exports[`The stack transformation can stack values 3`] = `
Array [
  Object {
    "c": 0,
    "x": 0,
    "y": 28,
    "y0": 35,
    "y1": 63,
    Symbol(vega_id): 1,
  },
  Object {
    "c": 1,
    "x": 0,
    "y": 55,
    "y0": 63,
    "y1": 118,
    Symbol(vega_id): 2,
  },
  Object {
    "c": 0,
    "x": 1,
    "y": 43,
    "y0": 9.5,
    "y1": 52.5,
    Symbol(vega_id): 3,
  },
  Object {
    "c": 1,
    "x": 1,
    "y": 91,
    "y0": 52.5,
    "y1": 143.5,
    Symbol(vega_id): 4,
  },
  Object {
    "c": 0,
    "x": 2,
    "y": 81,
    "y0": 9.5,
    "y1": 90.5,
    Symbol(vega_id): 5,
  },
  Object {
    "c": 1,
    "x": 2,
    "y": 53,
    "y0": 90.5,
    "y1": 143.5,
    Symbol(vega_id): 6,
  },
  Object {
    "c": 0,
    "x": 3,
    "y": 19,
    "y0": 23.5,
    "y1": 42.5,
    Symbol(vega_id): 7,
  },
  Object {
    "c": 1,
    "x": 3,
    "y": 87,
    "y0": 42.5,
    "y1": 129.5,
    Symbol(vega_id): 8,
  },
  Object {
    "c": 0,
    "x": 4,
    "y": 52,
    "y0": 26.5,
    "y1": 78.5,
    Symbol(vega_id): 9,
  },
  Object {
    "c": 1,
    "x": 4,
    "y": 48,
    "y0": 78.5,
    "y1": 126.5,
    Symbol(vega_id): 10,
  },
  Object {
    "c": 0,
    "x": 5,
    "y": 24,
    "y0": 40,
    "y1": 64,
    Symbol(vega_id): 11,
  },
  Object {
    "c": 1,
    "x": 5,
    "y": 49,
    "y0": 64,
    "y1": 113,
    Symbol(vega_id): 12,
  },
  Object {
    "c": 0,
    "x": 6,
    "y": 87,
    "y0": 0,
    "y1": 87,
    Symbol(vega_id): 13,
  },
  Object {
    "c": 1,
    "x": 6,
    "y": 66,
    "y0": 87,
    "y1": 153,
    Symbol(vega_id): 14,
  },
  Object {
    "c": 0,
    "x": 7,
    "y": 17,
    "y0": 54.5,
    "y1": 71.5,
    Symbol(vega_id): 15,
  },
  Object {
    "c": 1,
    "x": 7,
    "y": 27,
    "y0": 71.5,
    "y1": 98.5,
    Symbol(vega_id): 16,
  },
  Object {
    "c": 0,
    "x": 8,
    "y": 68,
    "y0": 34.5,
    "y1": 102.5,
    Symbol(vega_id): 17,
  },
  Object {
    "c": 1,
    "x": 8,
    "y": 16,
    "y0": 102.5,
    "y1": 118.5,
    Symbol(vega_id): 18,
  },
  Object {
    "c": 0,
    "x": 9,
    "y": 49,
    "y0": 44.5,
    "y1": 93.5,
    Symbol(vega_id): 19,
  },
  Object {
    "c": 1,
    "x": 9,
    "y": 15,
    "y0": 93.5,
    "y1": 108.5,
    Symbol(vega_id): 20,
  },
]
`;
